# Googλe Pagerank Algo

In [1]:
import numpy as np
import pandas as pd 
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/img-pages/pages.png


# We begin our journey with 5 sites linked together as shown below that we want to rank by their importance

![Websites](https://cdn.discordapp.com/attachments/1138429642211086346/1159085935712022538/image_2023-10-04_131225064-removebg-preview.png?ex=651e9a65&is=651d48e5&hm=2b6880d3d4a8c3ceccfcec40ec213ab7e63a6f3dc9a721af806f73e8af7bd0ad&)

A=1/2

B=1/4

C=1

D=1/2

E=1/3


$$A= \begin{pmatrix}
0 &   1/4 & 1& 1/2 & 1/3\\
1/2&  0 &   0 &  0 &      1/3\\
1/2&  1/4 & 0 &  1/2 &   0\\
0 &   1/4 & 0 &  0        &1/3\\
0 &   1/4 & 0 & 0 & 0\\
\end{pmatrix}$$

# At first the sites will be ranked equally like so with a vector r
$$r_0 = \begin{pmatrix}
.2\\
.2\\
.2\\
.2\\
.2\\
\end{pmatrix}$$

# We'll then derive a transition matrix A based on the probabilies of the 5 websites


$$A= \begin{pmatrix}
0 &  1/4 & 0 & 1/3\\
1/2& 0 & 0 & 1/3\\
1/2& 1/4 & 0 & 0\\
0 & 1/4 & 1 &1/3\\
0 & 1/4 & 0 & 0\\
\end{pmatrix}$$

In [2]:
# transition matrix
A = np.array([
[0 ,   1/4 , 1, 1/2 , 1/3],
[1/2,  0 ,   0 ,  0 ,      1/3],
[1/2,  1/4 , 0 ,  1/2 ,   0],
[0 ,   1/4 , 0 ,  0        ,1/3],
[0 ,   1/4 , 0 , 0 , 0]
])

r0 = np.array([.2, .2, .2, .2, .2])

A, r0

(array([[0.        , 0.25      , 1.        , 0.5       , 0.33333333],
        [0.5       , 0.        , 0.        , 0.        , 0.33333333],
        [0.5       , 0.25      , 0.        , 0.5       , 0.        ],
        [0.        , 0.25      , 0.        , 0.        , 0.33333333],
        [0.        , 0.25      , 0.        , 0.        , 0.        ]]),
 array([0.2, 0.2, 0.2, 0.2, 0.2]))

# Then to get the rankings you multiply r with A

$$r_1 = Ar_0$$
$$r_2 = Ar_1 = AAr_0 = A^2r_0$$
$$r_3 =  A^3r_0$$

and so on until
$$r_{\infty} = A^{\infty}v_0$$

In [3]:
r1 = np.matmul(A, r0)
r2 = np.matmul(A, r1)
# ....
# r_inf = np.matmul(A, r_inf-1) doesnt work ofc dont try it :trolled:

# How would we get $A^{\infty}$ you ask
# wit diagonalisationn

Like so
$$A = S \Lambda S^{-1}$$
which would give us
$$A^{2} = S \Lambda (S^{-1} S) \Lambda S^{-1} = S \Lambda^{2} S^-1$$

so
$$A^{\infty} = S \Lambda^{\infty} S^{-1}$$

and we have to do is get $Λ^{\infty}$

In [4]:
# We then get our eigenvalues and eigenvectors
λ,S = np.linalg.eig(A) 
S_inv = np.linalg.inv(S)

print("eigenvalues:", λ, "\n")
print("eigenvector matrix", S, "\n")
print("eigenvector matrix inverse", S_inv)

eigenvalues: [ 1.        +0.j         -0.55885384+0.j         -0.5       +0.j
  0.02942692+0.27146163j  0.02942692-0.27146163j] 

eigenvector matrix [[-0.72969394+0.j         -0.56490674+0.j         -0.49236596+0.j
   0.34467786+0.30135262j  0.34467786-0.30135262j]
 [-0.39801488+0.j          0.68934929+0.j          0.73854895+0.j
   0.26028638-0.30135262j  0.26028638+0.30135262j]
 [-0.5306865 +0.j          0.30837638+0.j          0.24618298+0.j
   0.24862149+0.26665927j  0.24862149-0.26665927j]
 [-0.13267163+0.j         -0.12444255+0.j         -0.12309149+0.j
  -0.60496425+0.j         -0.60496425-0.j        ]
 [-0.09950372+0.j         -0.30837638+0.j         -0.36927447+0.j
  -0.24862149-0.26665927j -0.24862149+0.26665927j]] 

eigenvector matrix inverse [[-5.28940822e-01-4.82303680e-18j -5.28940822e-01+8.26504749e-17j
  -5.28940822e-01-4.91291966e-19j -5.28940822e-01+9.28545370e-17j
  -5.28940822e-01+5.87674154e-18j]
 [-3.28729204e+00+4.35618724e-17j -2.20797202e+00+2.89196924e-16j
   

In [5]:
Λ = np.diag(λ)
Λ

array([[ 1.        +0.j        ,  0.        +0.j        ,
         0.        +0.j        ,  0.        +0.j        ,
         0.        +0.j        ],
       [ 0.        +0.j        , -0.55885384+0.j        ,
         0.        +0.j        ,  0.        +0.j        ,
         0.        +0.j        ],
       [ 0.        +0.j        ,  0.        +0.j        ,
        -0.5       +0.j        ,  0.        +0.j        ,
         0.        +0.j        ],
       [ 0.        +0.j        ,  0.        +0.j        ,
         0.        +0.j        ,  0.02942692+0.27146163j,
         0.        +0.j        ],
       [ 0.        +0.j        ,  0.        +0.j        ,
         0.        +0.j        ,  0.        +0.j        ,
         0.02942692-0.27146163j]])

In [6]:
# Now to test if its diagonalisable to begin with :monke:
np.linalg.det(S) # niccc det≠0

(-7.72320103186413e-18-0.06956441057842146j)

In [7]:
# and final test
def round(values, decs=0): 
  return np.round(values*10**decs)/(10**decs)


A_test = np.dot(S,np.dot(Λ, S_inv))


print("Expecting:", A, "\n")
print("Got :", round(A_test, decs=10)) # lesssgooooooo

Expecting: [[0.         0.25       1.         0.5        0.33333333]
 [0.5        0.         0.         0.         0.33333333]
 [0.5        0.25       0.         0.5        0.        ]
 [0.         0.25       0.         0.         0.33333333]
 [0.         0.25       0.         0.         0.        ]] 

Got : [[ 0.        +0.j  0.25      -0.j  1.        +0.j  0.5       -0.j
   0.33333333-0.j]
 [ 0.5       +0.j  0.        -0.j -0.        +0.j -0.        +0.j
   0.33333333+0.j]
 [ 0.5       +0.j  0.25      -0.j  0.        +0.j  0.5       +0.j
  -0.        +0.j]
 [ 0.        +0.j  0.25      -0.j -0.        +0.j -0.        +0.j
   0.33333333+0.j]
 [-0.        +0.j  0.25      -0.j  0.        +0.j  0.        +0.j
  -0.        +0.j]]


from this we get
$$Λ = \begin{pmatrix}
1.  &  0 & 0 &  0 &0\\
 & -0.55885384+0.j & 0 &  0 & 0\\
 &  0 & -0.5+0.j &  0 & 0 \\
0 &  0 & 0 &  0.02942692+0.27146163j & 0 \\
0 &  0 & 0 &  0 & 0.02942692-0.27146163j \\
\end{pmatrix}$$

and as we approach infinity the other eigenvalues approach zero like so
from this we get
$$Λ^{\infty} = \begin{pmatrix}
1.  &  0 & 0 &  0 &0\\
0 & 0 & 0 &  0 & 0\\
0 &  0 & 0 &  0 & 0 \\
0 &  0 & 0 &  0 & 0 \\
0 &  0 & 0 &  0 & 0\\
\end{pmatrix}$$

recall 
$$r_{\infty} = A^{\infty}v_0$$
$$= S Λ^{\infty}S^{-1}v_0$$

In [8]:
Λ_inf = np.array([
[1.  ,  0 , 0 ,  0 ,0],
[0 , 0 , 0 ,  0 , 0],
[0 ,  0 , 0 ,  0 , 0 ],
[0 ,  0 , 0 ,  0 , 0 ],
[0 ,  0 , 0 ,  0 , 0]]
)

In [9]:
r_inf = S@Λ_inf@S_inv@r0


# Final Ranking is as follows

In [10]:
sort_order= np.argsort(r_inf)+1
order = [str(index) for index in sort_order]
print(" < ".join(order))

5 < 4 < 2 < 3 < 1
